### ISSUE TITLE
QueryIter implements ExactSizeIterator, but doesnt provide an exact size_hint

### ISSUE THREAD

**Issue by MinerSebas** [bevyengine/bevy#1686]
## Bevy version

cd4c684ad5563c99b01e4ee6eea7c9b560c5dd5a

## What you did

I read the documentation of [ExactSizeIterator](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html), and notice that this requirement isn't true:
```
When implementing an ExactSizeIterator, you must also implement Iterator.
When doing so, the implementation of Iterator::size_hint must return the exact size of the iterator.
```
In the [standard implementation](https://doc.rust-lang.org/src/core/iter/traits/exact_size.rs.html#96-104) this is checked with an assertion.

Instead of providing an exact size_hint, Bevy [overrides](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/query/iter.rs#L119-L125) that function, which removes the need for that invariant to be true.

While it isn't a problem here, it would still be nice if QueryIter would provide an exact size_hint if it can.

**Comment by mockersf** [comment]
This may not be possible currently in rust without trait specialisation: https://github.com/rust-lang/rust/issues/31844

`size_hint` comes from `Iter` impl, which is:
https://github.com/bevyengine/bevy/blob/cd4c684ad5563c99b01e4ee6eea7c9b560c5dd5a/crates/bevy_ecs/src/query/iter.rs#L48-L50

but `ExactSizeIterator ` is
https://github.com/bevyengine/bevy/blob/cd4c684ad5563c99b01e4ee6eea7c9b560c5dd5a/crates/bevy_ecs/src/query/iter.rs#L118

so without a specialisation of `Iterator for QueryIter<'w, 's, Q, F>` we can't have a different `size_hint` for that case.

**Comment by MinerSebas** [comment]
Playing around I got this almost to work:
```rust
    fn size_hint(&self) -> (usize, Option<usize>) {
        if TypeId::of::<F>() == TypeId::of::<()>() {
            let temp = self
                .query_state
                .matched_archetypes
                .ones()
                .map(|index| self.world.archetypes[ArchetypeId::new(index)].len())
                .sum();

            return (temp, Some(temp));
        }

        (0, None)
    }
```
But then F needs to be `'static` and naively adding that constraint creates even more errors.

**Comment by cart** [comment]
If we can't meet the ExactSizeIter requirements for all cases, maybe we're better of removing ExactSizeIterator in favor of a manual len() impl for unfiltered queries.

**Comment by james7132** [comment]
This was resolved in #4244. It's currently inaccurate for partially iterated iterators, which is documented in #5149 with a fix in #5214. Closing this.
